### ISSUE TITLE
Formatter problem with multi-line strings

### ISSUE THREAD

**Issue by jakepearson** [BetterThanTomorrow/calva#2270]
I was writing a multi-line `SQL` as a `def` and found an interesting issue. I have turned off all extensions except for `Calva` and removed the contents of my `cljfmt.edn` to simplify and it still happens. I wasn't able to repro this by calling `cljfmt` from a prompt.

To repro, start with a file like

```clojure
(ns a-project.repro)

(def a-def "
            
            ")

(defn a-function [])
```

Put your cursor on the end of the blank line after `a-def`. Type a letter and then format the file and it turns into

```clojure
(ns little-differ.repro)

(def a-def "
a
")

(defn a-function [])"
```

Notice that an extra `"` is added to the end of the file after the last `)`. If I take out the newlines at the beginning and end of my string then no problems happen. Any ideas? Happy to work on a fix if people have ideas what I should do.

**Comment by bpringe** [comment]
> Notice that an extra " is added to the end of the file

I saw this at least once or twice recently and wasn't sure what caused it! Thanks for the repro.

Pinging @PEZ.

**Comment by PEZ** [comment]
Thanks for the report!

I can reproduce it with this document:

```clojure
 "
"
```

But only if I update the first line of the string. So it the above was the start document, and I update it such:

```clojure
 "a
"
```

Then I get the extra quote, and also the content of the string gets updated, which might be a clue:

```clojure
 " a
""
```

Then if I undo to get:

```clojure
 "a
"
```

And format the document, nothing happens. Not even if I add something to get formatted do I get a repro of the bug:

```clojure
 "
"
( a)
```

Becomes 

```clojure
 "
"
(a)
```

As it should. This is significant, because Calva checks the formatter result before applying any edits.

It all gets a bit extra crazy if we instead of **Format Document**, use **Format Selection**.  If I have:

```clojure
   "
"
```

Then select the string, and format selection, I get:

```clojure
   "
   "
```

If I keep the selection and format again, I get:

```clojure
   "
      "
```

So it seems it is the indent of the opening quote that gets inserted. It gets inserted on all lines in the string, following the first:

```clojure
   "a
a
"
```

**Format Selection** a few times:


```clojure
   "a
               a
               "
```

Updating the first line:

```clojure
   "aa
               a
               "
```

and then selecting the string and format the selection:

```clojure
   " aa
   a
   ""
```

Maybe there are two different bugs in play, even.

You are super welcome to have a look at this and what the potential fix might be, @jakepearson. Since we get different behaviour with the OP issue depending on wether the test is updated or not I suspect the bug is somewhere in the mutable TypeScript land. I also suspect it will be a bit of work to pinpoint the root cause.

The second issue/symptom is easier to reproduce without mutation, so maybe try to expose it with some tests in `src/cljs-lib/test/calva/fmt/formatter_test.cljs`, because this issue might be in the CLJS code.

Neither the TS, nor the CLJS code for all this is in very tidy shape, so if it all gets too messy to try to disentangle, just let me know and I could possibly shed some light on things. Or pick things up wherever you run out of stamina, should that happen. 😄

**Comment by PEZ** [comment]
As for further ideas of what to do, @jakepearson, I think it will be about using breakpoints and log-points in `format.ts` to try catch where the shenanigans starts. It's truly strange with this that it the same text gets differently treated depending on if it's recently updated or not. Strange to the point where it is slightly worrisome, I would say.

**Comment by jakepearson** [comment]
Thanks for the starting ideas. I'll take some time next week (hackathon week).

**Comment by bpringe** [comment]
It might also be a good idea to go back to commits of recent releases and see if you can reproduce the issue there. If you can't, then you have a smallish change set to provide clues on where the issue lies.

**Comment by bpringe** [comment]
I say the above partly because I hadn't seen a double quote be added to a file until only recently.

**Comment by jakepearson** [comment]
@bpringe Good call. It for sure seems like a recent issue.

**Comment by jakepearson** [comment]
Hi. I bisected and landed on this commit as the issue (https://github.com/BetterThanTomorrow/calva/commit/d5f585afdbfffba2997a9bf9eb539e2d3b07017f). I'll see if I can fiddle around with it a bit and get a fix. The lamest solution was a revert, but there was a conflict over the whole section.

**Comment by jakepearson** [comment]
I might have hit where I can be useful with my very limited knowledge of how Calva works. Here is what I have learned:
- The value on this line (https://github.com/BetterThanTomorrow/calva/blob/published/src/calva-fmt/src/format.ts#L59) for `cursor.withinString()` is `false` when formatting a file that is working incorrectly. I assume this should be `true`.
- With a breakpoint here (https://github.com/BetterThanTomorrow/calva/blob/published/src/cursor-doc/token-cursor.ts#L830) the cursor type is `eol`. I added `|| cursor.getPrevToken().type === 'eol'` to the if statement and my bug goes away (but makes a bunch of tests in `cursor-token-test.ts` fail). I assume this would be a bad fix and cause other issues.
- I added a test to `cursor-token-test.ts` to try to repro the bug, but those appear to pass as we would expect.

``` clojure
it('is true in multiline string', () => {
  const contexts = context.determineContexts(
    docFromTextNotation('"•        abc|•        def•       •       "')
  );
  expect(contexts.includes('calva:cursorInString')).toBe(true);
});
```

Please let me know what the next steps should be for me or if you would like to take over.

Have a great day!

**Comment by PEZ** [comment]
Sorry about the radio silence, @jakepearson! I had completely missed that you had come back with these results. Let me read what you have written and think a bit and I will try to figure what the next steps might be.

**Comment by PEZ** [comment]
One thing that this reminds me about is that we have a one or two tests involving strings that are disabled because they cause other tests to fail. I think that the lexical scanner might have a leak. No idea if that is related to this bug, but it could be...

**Comment by PEZ** [comment]
> but makes a bunch of tests in cursor-token-test.ts fail

Do these tests involve `withinString`, or do they seem unrelated to that?

**Comment by jakepearson** [comment]
It was last week (so my memory is mostly gone), but I think so.

**Comment by PEZ** [comment]
I'll make the same experiment and see what I get.

**Comment by jakepearson** [comment]
@PEZ I was wondering if you had made any progress on this issue or if there was anything I could help with?

**Comment by PEZ** [comment]
Oh, I started from the end of cljfmt, trying to eliminate it as the source. Ended up upgrading it and wrecking havoc with peoples formatting configs... Then I forgot about why I had started it all. I did add a test or two trying to expose this, but also failed. And added a test file.

I'm a bit unsure what the next step is, tbh. Probably chasing the bug with breakpoints in `format.ts` and that test file I added: https://github.com/BetterThanTomorrow/calva/commit/061c1a0ce512e32a90090f42e5687e3884254379

**Comment by evaogbe** [comment]
I'm also running into the same issue

## Steps to Reproduce

1. Create a file containing a multiline string literal
```clojure
(def single-line-str
  "Ut labore et dolore")

(def multi-line-str
  "Lorem ipsum dolor sit amet
   consectetur adipiscing elit
   sed do eiusmod tempor incididunt")

(defn inc-fn
  [x]
  (
   inc x))
```
2. Edit any line of the multiline string except the last one
    For example, add `NEW TEXT` to the second line of `multi-line-str`
3. Run the "Format Document" command

## Expected Outcome

The file is formatted (if formatting needed) and syntax remains valid

```clojure
(def single-line-str
  "Ut labore et dolore")

(def multi-line-str
  "Lorem ipsum dolor sit amet
   consectetur adipiscing elit NEW TEXT
   sed do eiusmod tempor incididunt")

(defn inc-fn
  [x]
  (inc x))
```

## Actual Outcome

The file is mangled:
- Does not format the file
- Adds a space to the front and end of every string literal in the file
- Removes all whitespace from the beginning of every line of the multiline string
- Adds a single quotation mark to the end of the file
- The file is no longer valid Clojure syntax because of the unclosed ending quotation mark. (clj-kondo reports "Unexpected EOF while reading string.")

```clojure
(def single-line-str
  " Ut labore et dolore ")

(def multi-line-str
  " Lorem ipsum dolor sit amet
consectetur adipiscing elit NEW TEXT
sed do eiusmod tempor incididunt ")

(defn inc-fn
  [x]
  (
   inc x))"
```

It mangles the file even if it's already formatted correctly and should not be changed at all. However, it doesn't mangle the file if the last change before formatting is not an edit to the multiline string.

I also tried adding a single unclosed quotation mark to the end of the file before formatting. It doesn't add a second quotation mark to the end of the file but it still does all the other mangling. It will add another unclosed quotation mark to the end of the file if I end the file with an empty string. So this bug will always make the file have invalid syntax.

Running the formatter on the command line (i.e. `clojure-lsp format` or `clj -Tcljfmt fix`) correctly produces the expected outcome.

**Comment by bpringe** [comment]
Thanks for the additional info @eoogbe!

**Comment by bpringe** [comment]
I did a little bit of looking into this so far. With the following text in a file:

```clojure
(ns calva-issue-2270
  "To reproduce the issue, update the multi-line-str docstring then run the command Format Document.")

(def multi-line-str
  "Lorem ipsum dolor sit amet
   consectetur adipiscing elit
   sed do eiusmod tempor incididunt")

(defn inc-fn
  [x]
  (inc x))

```

If I put the cursor in the docstring of `multi-line-str` and format the document, at [this line](https://github.com/BetterThanTomorrow/calva/blob/f6ec93c2b310751bff41f66f81bb549d2ec0bf03/src/calva-fmt/src/format.ts#L64), `missingTexts` is:

```javascript
{
  prepend: "\"",
  append: "\"",
}
```

which I find odd. Then `healedText` is:

```javascript
'"(ns calva-issue-2270\n  "To reproduce the issue, update the multi-line-str docstring then run the command Format Document.")\n\n(def multi-line-str\n  " Lorem ipsum dolor sit amet\n   consectetur adipiscing elit\n   sed do eiusmod tempor incididunt")\n\n(defn inc-fn\n  [x]\n  (inc x))"'
```

Notice the original text of the document is now wrapped in double quotes. This then gets passed to `formatCode`, which returns a promise that resolves to the undesirable new document text:

```javascript
'"(ns calva-issue-2270\n  " To reproduce the issue, update the multi-line-str docstring then run the command Format Document. ")\n\n(def multi-line-str\n  " Lorem ipsum dolor sit amet\nconsectetur adipiscing elit\nsed do eiusmod tempor incididunt ")\n\n(defn inc-fn\n  [x]\n  (inc x))"'
```

That text is later transformed in a way that the `"` at the beginning of the file is removed, but the one at the end remains (see the value of `formattedText` while debugging).

**Comment by bpringe** [comment]
The output of `getMissingBrackets` [here](https://github.com/BetterThanTomorrow/calva/blob/f6ec93c2b310751bff41f66f81bb549d2ec0bf03/src/calva-fmt/src/format.ts#L64) is different depending on whether the multi-line string was the last thing edited before formatting, even given the same exact string argument (the contents of the file), which means `getMissingBrackets` is stateful in a way that is causing undesirable returned data in the issue case. What it returns depends on whether the multi-line string was the last thing edited before it was run.

There may be other issues besides that, but I think starting with fixing the issue with `getMissingBrackets` would be wise.

**Comment by bpringe** [comment]
@PEZ I don't know the history of the formatting code in Calva, but it seems that it does things in addition to what cljfmt does to format the text. Can you summarize what additional things Calva does and why? I'm wondering if it would make sense to  get rid of the additional things (or some of them) and just use cljfmt as purely as possible.

Maybe some of those additional things were added to compensate for things cljfmt didn't do or wasn't good at in the past, but can handle now? I'm guessing also maybe some of the additional code is to handle VS Code details.

**Comment by PEZ** [comment]
@bpringe We do no additional formatting over what cljfmt does. There are some manipulation of the input we give cljfmt, which we do to:

* prevent folding of brackets
* prevent trimming of the empty line where the cursor is
* get a new cursor position for the cursor on empty lines (we use this when the user hits enter, or presses tab on an empty line)

This is also partly achieved by relaxing some cljfmt settings when we perform on-type formatting (while editing, as opposed to direct formatting commands). We can get rid of some of this by abandoning the legacy indenter in favor of the “new” one. The problem with that is that the new one has some issues that people get around by switching to the old one. Fixing those issues would take us far. I have improved it quite a lot the past few months, but it is still misbehaving in some situations.

I'll have a look at your latest findings. Sounds a lot on the descriptions here like it has to do with the bug in the token scanner that has us disable some unit tests. Enabling these causes a lot of following unit tests to fail. Some statefulness there that we haven't figured out, but obviously need to figure out.

**Comment by PEZ** [comment]
Now had a bit of a look at your findings, @bpringe, especially this line:

https://github.com/BetterThanTomorrow/calva/blob/f6ec93c2b310751bff41f66f81bb549d2ec0bf03/src/calva-fmt/src/format.ts#L64

With some luck the whole issue is less hairy than the ugly stateful problem we have in the token scanner. That line was added by me on June 25 as part of making unbalanced text format:

https://github.com/BetterThanTomorrow/calva/blob/f3bca8405a4dbc7fb79a03207a65762c008116c3/CHANGELOG.md#L171-L174

Though, since you notice different behaviour of `getMissingBrackets` depending on some editor state, I fear we are actually running up against this unsolved problem with the scanner going crazy affecting future runs...

Seeing that this issue was filed a few days later, I now think it is a regression and some oversight on my part, maybe around how string quotes are also brackets. Maybe we can reproduce the error with some assertions here:

https://github.com/BetterThanTomorrow/calva/blob/a434f6537d0d75afc0fe5d5e082954d05ffe8b43/src/extension-test/unit/cursor-doc/utilities-test.ts#L6-L36

**Comment by bpringe** [comment]
Thanks for the info @PEZ.

> Maybe we can reproduce the error with some assertions here

I haven't done so yet, but I assume the tests will pass as expected since the issue seems to depend on an edit happening in a multi-line string prior to that function being run. Maybe we can reproduce _that_ in a test, though. I'm not sure if that would involve an integration test, but maybe an integration test is a good idea.

Edit: I did add a unit test for it and it passed as expected.

**Comment by bpringe** [comment]
Just to add some more info:

If I hardcode `missingTexts` [here](https://github.com/BetterThanTomorrow/calva/blob/f6ec93c2b310751bff41f66f81bb549d2ec0bf03/src/calva-fmt/src/format.ts#L64) to be `{ prepend: '', append: '' }`, then run the repro case, the issue does not occur.

That leads me to believe the issue may be fully or mostly isolated to `getMissingBrackets`.

**Comment by jakepearson** [comment]
Yay! Thanks so much.
